Carbon.Security/Functions/Unprotect-CString.ps1

```diff
@@ -0,0 +1,323 @@
+                return
+            }
+
+            if( -not $Certificate.PrivateKey )
+            {
+                $msg = "$($certDesc) has a private key, but it is null or not set. This usually means your certificate " +
+                    'was imported incorrectly or was created without a private key. Make sure you''ve generated an ' +
+                    'RSA public/private key pair and are using the private key. If the private key is in the Windows ' +
+                    'certificate store, make sure the current user has permission to read the private key (use ' +
+                    'Carbon''s `Grant-CPermission` function).'
+                Write-Error -Message $msg -ErrorAction $ErrorActionPreference
+                return
+            }
+
+            [Security.Cryptography.RSA]$privateKey = $null
+            $privateKeyType = $Certificate.PrivateKey.GetType()
+            $isRsa = $privateKeyType.IsSubclassOf([Security.Cryptography.RSA]) 
+            if( -not $isRsa )
+            {
+                $msg = "$($certDesc) is not an RSA key. Found a private key of type ""$($privateKeyType.FullName)"", but " +
+                    "expected type ""$([Security.Cryptography.RSA].FullName)"" or one of its sub-types."
+                Write-Error -Message $msg -ErrorAction $ErrorActionPreference 
+                return
+            }
+
+            if( -not $Padding )
+            {
+                $Padding = [Security.Cryptography.RSAEncryptionPadding]::OaepSHA1
+            }
+
+            $privateKey = $Certificate.PrivateKey
+            $decryptedBytes = $privateKey.Decrypt($encryptedBytes, $padding)
+        }
+        elseif( $PSCmdlet.ParameterSetName -eq 'Symmetric' )
+        {
+            $Key = ConvertTo-AesKey -InputObject $Key -From 'Unprotect-CString'
+            if( -not $Key )
+            {
+                return
+            }
+
+            $aes = [Security.Cryptography.Aes]::Create()
+            try
+            {
+                $aes.Padding = [Security.Cryptography.PaddingMode]::PKCS7
+                $aes.KeySize = $Key.Length * 8
+                $aes.Key = $Key
+                $iv = [byte[]]::New($aes.IV.Length)
+                [Array]::Copy($encryptedBytes, $iv, 16)
+
+                $encryptedBytes = $encryptedBytes[16..($encryptedBytes.Length - 1)]
+                $encryptedStream = New-Object -TypeName 'IO.MemoryStream' -ArgumentList (,$encryptedBytes)
+                try
+                {
+                    $decryptor = $aes.CreateDecryptor($aes.Key, $iv)
+                    try
+                    {
+                        $mode = [Security.Cryptography.CryptoStreamMode]::Read
+                        $cryptoStream = [Security.Cryptography.CryptoStream]::New($encryptedStream, $decryptor, $mode)
+                        try
+                        {
+                            $decryptedBytes = [Byte[]]::New($encryptedBytes.Length)

```

Comment:
If you create the byte array from the length of the encrypted data, you're including the padding bytes in the decrypted output. I fixed this in Carbon not too long ago: https://github.com/webmd-health-services/Carbon/blob/master/Carbon/Functions/Unprotect-String.ps1#L245-L253

Reply:
Is there a test to catch this situation? A known plaintext value that gets returned incorrectly?

Reply:
Yeah see the repro here: https://github.com/webmd-health-services/Carbon/issues/87
And the PR included a test that catches this scenario: https://github.com/webmd-health-services/Carbon/commit/aa8e79b71dc115fdce74b9e219079e7ffd715173#diff-e4e33fdc888396d3db0944dcc439490618ea46f15f9deb4040608f094e37972d

Reply:
Thanks. I was missing your test for some reason. Weird.